Implement DCGAN example with distributed support (DDP)

jafraustro opened the pull request:
- DDP is recommended instead of DataParallel https://docs.pytorch.org/docs/stable/generated/torch.nn.DataParallel.html
- Fix original DCGAN example, `DataParallel` does not suport xpu devices

gera-aldama commented on distributed/dcgan/main.py:
```diff
@@ -0,0 +1,329 @@
+        import sys
+        sys.exit()
+
+    parser = argparse.ArgumentParser(
```
Shouldn't this part of the code be in `main()` instead of it being empty?

jafraustro replied:
Thanks, I will fix it

gera-aldama commented on dcgan/main.py:
```diff
@@ -192,7 +192,7 @@ def __init__(self, ngpu):
 
     def forward(self, input):
-        if (input.is_cuda or input.is_xpu) and self.ngpu > 1:
+        if (input.is_cuda) and self.ngpu > 1:
```
Can we have something different here so that it doesn't only focus on cuda?

`if torch.accelerator.is_available()`?

`if not input.device.type == 'cpu'`?

jafraustro replied:
Thanks for the feedback.

`data_parallel` only works for cuda.

gera-aldama reviewed (commented):
Overall LGTM, just added two comments.

eromomon reviewed (approved):
LGTM

jafraustro committed:
Implement DCGAN example with distributed support (DDP)

- DDP is recommended instead of DataParallel https://docs.pytorch.org/docs/stable/generated/torch.nn.DataParallel.html
- Fix original DCGAN example, `DataParallel` does not suport xpu devices
- Improve code readability

Signed-off-by: jafraustro <jaime.fraustro.valdez@intel.com>

jafraustro commented:
PR to upstream
